#DEGs (0.83 > FC > 1.2 and q < 0.01 or q < 0.05) were investigated for enrichment 
#in the curated pathway categories “Cell Process” and “Signal Processing” 
#of the Pathway Studio software (Elsevier) version  12.4.0.5 by using Fisher’s exact test.
#Subsequently, p-value correction for multiple testing with the fdrtool program (v1.2.16) 
#was performed. 

library(fdrtool)

#Example: NSC1a - Signal Processing; DEGs q < 0.05
signal_pathways.nsc1a <- c("JNK/MAPK Signaling","TNFRSF6 -> HSF1 Signaling","ERK5/MAPK7 Signaling","mTOR Signaling Activation by Fatty Acids and Glucose","EphrinB -> MAPK/JUN/FOS Signaling","EphrinB -> Cytoskeleton Signaling","MTNR1 Signaling","EctodysplasinR -> AP-1 Signaling","NOTCH -> TCF3 Signaling","P38 MAPK/MAPK14 Signaling","FPR1 -> Cytoskeleton Signaling","ERK/MAPK Canonical Signaling","ADRA2C/ADRB2 -> Synaptic Endocytosis","IL6ST -> STAT5B Signaling","ADORA3 -> Mast-Cell Degranulation","HTR4/6/7 -> Cation Channels","IGF1R -> ELK/SRF/HIF1A/MYC/SREBF Signaling","IL6R -> CEBP/ELK/SRF Signaling","ADORA1/2A -> Exocytosis","ADRA2C/ADRB2 -> Vasoconstriction","NGFR -> AP-1/CEBPB/CREB/ELK/SRF/TP53 Signaling","PTGER1/4 -> Vascular Motility","TNFR -> AP-1/ATF/TP53 Signaling","TNFRSF6 -> FOXO3A Signaling","Ras-GAP Regulation Signaling","IFNG -> ARRB1/STAT1 Signaling","PTPRJ -> CTNND Signaling","PTPRU -> CTNNB Signaling","Sialophorin -> CTNNB/MYC/TP53 Signaling","EGFR/ERBB -> STAT Signaling","EDNRA/B -> Vascular Motility","EGFR/ERBB2 -> CTNNB Signaling","mTOR Signaling","PTGIR -> Vasodilation","ADRA1 -> Vasoconstriction","TNFR -> CREB/ELK-SRF Signaling","FibronectinR -> AP-1/ELK/SRF/SREBF Signaling","TGFBR -> ATF/GADD/MAX/TP53 Signaling","TGFBR1/AP-1 Signaling","CHRM1/2/3 -> Vascular Motility","NCAM1 -> CREB/ELK/SRF/MYC Signaling","Ubiquitin Ligase Complex in Prostata Cell","TNFRSF1A -> AP-1/ATF/TP53 Signaling","DRD1/5 Expression Targets","HRH1/3 -> Synaptic Transmission","OR1A1 -> GNAS Signaling","AGER -> CREB/SP1 Signaling","NOTCH -> RBPJ/HES/HEY Signaling","IL4R -> ELK/SRF/HMGY Signaling","VEGFR -> CTNND Signaling","TGFBR -> SMAD1/5/9 Signaling","IL1R -> STAT3 Signaling","FGFR1 -> STAT Signaling","NTRK -> AP-1/CREB/ELK/SRF/MYC/SMAD3/TP53 Signaling","CD157/ITGB2 Signaling in Myeloid Cell",
                           "IGF1R -> MEF/MYOD/MYOG Signaling","ActivinR/BMPR -> SMAD1/5/9 Signaling","TGFBR -> MEF/MYOD/MYOG Signaling","HGFR -> STAT Signaling","TNFRSF6 -> DDIT3 Signaling","TGFBR/BMPR -> SMAD2/3 Signaling","ARRB1/ARRB2 non-CanonicalSignaling and HedgehogFamily","Androgen Receptor Genomic Signaling","TNFRSF1A -> CREB/ELK-SRF Signaling","HRH1/2 -> Vascular Motility","ADRB1/3 -> Vasodilation","EphrinR -> STAT Signaling","EGFR -> CTNND Signaling","FibronectinR -> NF-kB Signaling","PDGFR -> STAT Signaling","IGF1R -> STAT Signaling","IL1R -> NF-kB Signaling","HTR4/6/7 -> Vasodilation","EGFR/ERBB2 -> HIF1A Signaling","mTOR Signaling Activation by Amino Acids","VEGFR -> ATF/CREB/ELK-SRF Signaling","FrizzledR -> JUN/PAX2 Signaling","EGFR -> AP-1/CREB/ELK/SRF/MYC Signaling","GDNF -> HSF1 Signaling","CGRP -> Calcium Influx","ADORA2A/B -> Vasodilation","InsulinR -> ELK/SRF/SREBF Signaling","TNFRSF1A -> STAT Signaling","NOTCH -> MEF/MYOD Signaling","VEGFR -> STAT Signaling","VEGFR -> CTNNB Signaling","TGFBR -> CREB/ELK-SRF Signaling","PECAM -> STAT Signaling","Frizzled Receptors -> ARRB1/ARRB2 Canonical Signaling","DRD2/4 -> Membrane Transport","Ras-GRF Regulation Signaling","PTGFR -> Vasoconstriction","HTR1 -> Vascular Motility","NOTCH Receptors Signaling","HIF1 Signaling","EGFR/ERBB2 -> TP53 Signaling","PTGDR -> Vasodilation","EGFR -> AP-1/ATF2 Signaling","TNFRSF5 -> STAT Signaling","IL11R -> STAT3 Signaling","KIT -> STAT Signaling","ALK -> STAT Signaling","PTGER2/3 -> Vascular Motility","EGFR -> ZNF259 Signaling","PDGFR -> AP-1/MYC Signaling","FibronectinR -> CTNNB Signaling","CHRNA7 -> CREB Signaling","FibronectinR -> ICAP-1A/MYC Signaling","Frizzled Receptors -> ARRB1/ARRB2 non-Canonical Signaling","FGFR -> AP-1/CREB/CREBBP/ELK/SRF/MYC Signaling","PECAM -> SP1 Signaling","HTR7 -> IL6 Production","FGFR3 -> STAT Signaling",
                           "NTRK -> FOXO/MYCN Signaling","CNR1/2 -> Vascular Motility","INSR -> ARRB2/AKT/SRC Signaling","IL13R -> STAT6 Signaling","IL4R -> STAT Signaling","CSF2 -> STAT Signaling","SELE -> ELK-SRF Signaling","AngiopoietinR -> AP-1 Signaling","GABA(B)R -> Postsynaptic Inhibition","ICAM1 -> AP-1/CREB/ELK/SRF Signaling","NGFR -> MEF Signaling","IGF1R -> ARRB1/ERK1/3 Signaling","UrokinaseR -> ELK/SRF Signaling","CD2 -> STAT Signaling","IL31R -> STAT Signaling","IL9R -> STAT Signaling","IL6R -> STAT Signaling","IL21R -> STAT Signaling","UrokinaseR -> STAT Signaling","TNFRSF5/6 -> RB1/E2F Signaling","TGFBR -> AP-1 Signaling","AHR Signaling in Th17 Cells Function","ICAM2 -> CTNNB/FOXO/STAT3 Signaling","HCAR1/HCAR2 -> MAPK Signaling","EGFR/ERBB3 -> MEF/MYOD/NFATC/MYOG Signaling","Oxidative Stress Protection via AHR-NFE2L2-NQO1 Signaling in Keratinocytes","InsulinR -> CTNNB/FOXA/FOXO Signaling","VEGFR -> AP-1/CREB/MYC Signaling","IL13R -> STAT Signaling","IL3R -> STAT Signaling","IL7R -> STAT Signaling","IL5R -> STAT Signaling","IL12R -> STAT Signaling","InsulinR -> STAT Signaling","ARRB2 and Frizzled Receptors Endocytosis","EctodysplasinR -> LEF1 Signaling","IL2R -> STAT Signaling","NOTCH -> SMAD3 Signaling","NOTCH -> LEF1 Signaling","ProlactinR -> STAT Signaling","EGFR -> NCOR2 Signaling","KIT -> MITF Signaling","HGFR -> AP-1/CREB/MYC Signaling","ERBB2/3 -> EP300/ETS/ETV/SP1 Signaling","CD19 -> JUN/ELK Signaling","IL2R -> ELK/SRF/MYC Signaling","IL6 Expression Targets -> Nociception","IL15R -> STAT Signaling","HTR2 -> Membrane Transport","EphrinR -> Actin Signaling","BDKRB1/2 -> Ion Channels","GPCRs Desentization","CHRNA7 -> NOS1 Production","HRH1/2 -> Membrane Polarization","NOTCH -> EP300/ASCL Signaling","MERTK Signaling","Leptin -> NO Production/Vasodilation","PTGER2/3 -> Inflammation-Related Expression Targets",
                           "FcIgER -> ELK-SRF Signaling","FrizzledR -> CTNNB Signaling","GRM2-4/6-8 (Presynaptic) -> Glutamate Release Attenuation","CD38/CD19 -> JUN/FOS/NF-kB Signaling in B-cell Proliferation","ADRA2A -> Hyperpolarization","ActivinR -> SMAD2/3 Signaling","PTPRC -> BCL6 Signaling","EPHB -> NMDA Receptor Activation","AHR Signaling in Tr1 Cells Function","T-Cell Receptor -> ATF/CREB Signaling","TNFR -> NF-kB Signaling","NF-kB non-Canonical Signaling","ADRB1 -> Prostaglandin Generation","ARRB1/ARRB2 and GPCRs Internalization","BDKRB1/2 -> Vasodilation","EGFR -> SMAD1 Signaling","Complement Component GPCRs Signaling","TACR1 -> Membrane Transport","AHR Signaling in Treg and Dendritic Cells Function","HRH2/4 -> IL6/10 Production","TNFRSF5/13B -> NFATC1 Signaling","IGF1R -> CEBPA/FOXO1A Signaling","AMPK Signaling","EctodysplasinR -> NF-kB Signaling","DDR1 -> NF-kB Signaling","ADRA1A -> IL6 Production","Kynurenine Metabolites in T-Cell Apoptosis","CNR1/2 -> IL1B/2/4/6/10 Production","CHRM1/2/3/5 -> Ion Channels","PTGIR -> IL6 Production","GUCYC2 Signaling","FGFR -> RUNX2 Signaling","ADRB1 -> Ion Channels","TAAR1 -> Neurotransmitter Uptake","GALR1/2/3 -> Neurotransmitter Metabolism","PI3K/RAC1 Signaling","IL1B Expression Targets -> Nociception","P2RY11/13/14 -> IL8/10 Production","DRD2 -> TRPC1 Transcription","CHRM1 -> IL2 Production","VEGFR -> NFATC Signaling","Hippo/YAP1 Signaling","OPRK -> Pain Perception","DRD1/3 -> Potassium Uptake","CNR1/2 -> Membrane Transport","TLR1/2/6 -> NF-kB Signaling","B-Cell Differentiation Inhibition by Dioxin","MacrophageR -> CEBPB -> NF-kB Signaling","ADRA1 -> Prostaglandin Generation","TLR -> AP-1 Signaling","B-Cell Receptor -> AP-1 Signaling","Kynurenine Metabolites in Neurotoxicity and Neuroprotection","DRD2 Expression Targets","OPRD/OPRM -> Ion Channels","T-Cell Receptor -> STAT Signaling",
                           "CD38/CD3 -> JUN/FOS/NF-kB Signaling in T-cell Proliferation","Estrogens/ESR1 Genomic Canonical Signaling","BDKRB1/2 -> Prostaglandin Generation","OPRL1 -> Ion Channels","NMDA Receptors -> Ca2+/CREB Activation/PGE2 Synthesis","GALR1/2/3 -> POMC/NPY Production","T-Cell Receptor -> AP-1 Signaling","AHR Signaling in M1 Macrophages Function","Leptin -> CD25/IL6/IL10 Production","GRM1/5 (Postsynaptic) -> Ion Channels","HTR1 -> IL6 Production","GLP1R -> ARRB1 Signaling","HTR1 -> Membrane Transport","T-Cell Receptor -> CREBBP Signaling","HRAS Signaling","Lysophosphatidic Acid/LPARs Signaling","Dioxin Induced Chloracne (Hypothesis)","AMPA Receptors -> Calcium Influx","ADRA1 -> Ion Channels","Aryl Hydrocarbon Receptor Genomic and non-Genomic Signaling","mTOR Signaling","TGF-beta Signaling","CD38 -> cADPR/Calcium Signaling","CRH -> Synthesis of Corticosteroids",
                           "CD157 -> cADPR/Calcium Signaling","WNT/Calcium non-Canonical Signaling","FOXO1 Signaling","Androgen Receptor non-Genomic Signaling","WNT Planar Cell Polarity (PCP) non-Canonical Signaling","HTR5 -> TNF Production","TACR1 -> TNF/IL6/IL8 Production","Ras Signaling","Insulin Signaling","TNF-alpha/TNFRSF1B Signaling","KRAS Signaling","Estrogens/ESR1 non-Genomic Signaling","NF-kB Canonical Signaling","Hedgehog Family Signaling","TGFBR3 -> ARRB2 Signaling","VEGF Signaling","TNF-alpha/TNFRSF1A Signaling","WNT Canonical Signaling","EGFR Signaling","AXL Receptor Tyrosine Kinase Signaling","AGTR1 -> ARRB1/ARRB2 Signaling","BMPR2 Signaling","Hedgehog Family -> ARRB1/ARRB2 Canonical Signaling")
pval.fisher <- c(2.01E-06,1.08E-04,1.65E-04,1.65E-04,2.15E-04,2.50E-04,5.99E-04,6.25E-04,7.57E-04,8.55E-04,9.04E-04,9.46E-04,0.001117783,0.001466184,0.001638203,0.001731892,0.002180907,0.002515162,0.002631335,0.002961238,0.00373928,0.004100179,0.004504845,0.004597114,0.006867313,0.007152757,0.007216279,0.007216279,0.007505594,0.007505594,0.008496339,0.008619956,0.008975759,0.009188183,0.009847327,0.010313554,0.010488008,0.011375961,0.011375961,0.0121795,0.012845967,0.014703841,0.015805797,0.015867118,0.015867118,0.015882005,0.016025265,0.016025265,0.016033389,0.016488116,0.016488116,0.019841142,0.019841142,0.02232697,0.02301512,0.02301512,0.023187553,0.023187553,0.024739187,0.025955618,0.025955618,0.026253101,0.026861717,0.029310799,0.029310799,0.029378839,0.03308033,0.03308033,0.033652932,0.034810224,0.034810224,0.037507403,0.037507403,0.039211378,0.041560929,0.042796117,0.043505894,0.049622444,0.050380283,0.054231781,0.055769908,0.058001963,0.060226383,0.060226383,0.060226383,0.060633589,0.060633589,0.060699633,0.061407199,0.062884172,0.07054755,0.07186438,0.075853668,0.076830577,0.076888521,0.079729479,0.083882495,0.085423755,0.086024377,0.086024377,0.086024377,0.086024377,0.091362786,0.09206839,0.094409724,0.094642979,0.094642979,0.09718441,0.099065255,0.105861046,0.106070094,0.108153699,0.110099146,0.111212474,0.112120766,0.113838882,0.113838882,0.113838882,0.113838882,0.118378447,0.120107419,0.12298163,0.124475384,0.126081213,0.130708154,0.143137656,0.14353878,0.14353878,0.14353878,0.14353878,0.14353878,0.14353878,0.149689249,0.149689249,0.154260843,0.158107232,0.166694149,0.17074305,0.170961057,0.170961057,0.173662649,0.174604081,0.174604081,0.174604081,0.174604081,0.174604081,0.174604081,0.18303097,0.206589246,0.206589246,0.206589246,0.206589246,0.206589246,0.218088691,0.21990648,0.226260109,0.229851773,0.232752973,
                 0.234466934,0.237526802,0.239114326,0.241835395,0.262621147,0.263693468,0.26640089,0.267022353,0.267814008,0.271857079,0.282867227,0.286527779,0.286614296,0.300226045,0.307146751,0.307146751,0.307336318,0.310594764,0.310594764,0.318148548,0.330537411,0.340073286,0.347013536,0.347013536,0.347013536,0.370218205,0.370324828,0.370324828,0.371358241,0.385522522,0.402738954,0.405196544,0.426219428,0.430772984,0.430772984,0.455916361,0.460491289,0.460491289,0.473382896,0.519933882,0.534876332,0.55367453,0.565301298,0.569617854,0.571534266,0.575523852,0.601921872,0.601921872,0.603547234,0.608052745,0.621701348,0.640412829,0.649106037,0.656643686,0.672904717,0.693758052,0.701724821,0.710293788,0.720140793,0.735642984,0.737590605,0.748245528,0.762965268,0.776799261,0.784458557,0.801099048,0.806823323,0.823816915,0.831605327,0.843884796,0.863877179,0.86697939,0.880110415,0.885011611,0.89792357,0.918865153,0.924319076,0.924992163,0.937949089,0.941563721,0.942715099,0.963871526,0.985770939,0.9881592,0.9902293,0.990769857,0.996024599,0.996782524,0.997210361,0.99742575,0.997835778,0.999053264,0.999475643,0.999559387,0.999594651,0.99974923,0.999830221,0.999848739,0.999963404,0.99998877,0.999990585,0.999997518,0.99999816,0.999999001,0.99999997,0.999999971,0.999999983,0.999999991,0.999999999,1,1,1,1,1,1)

pvalues.category <- matrix(pval.fisher)

uncorrected.pval <- pvalues.category[,1]

cutoff.method = "fndr"

fdr.out <- fdrtool(uncorrected.pval, statistic = "pvalue", cutoff.method = cutoff.method)

print(fdr.out$param[2] < 10) #if TRUE (N.cens < 10), i.e. small censored sample size, cutoff.method "pct0" is used  
                             #instead of "fndr" with fraction pct0 of data for fitting the null model

if(fdr.out$param[2] < 10) {
  fdr.out <- fdrtool(uncorrected.pval, statistic = "pvalue", cutoff.method = "pct0", pct0=pct0) #settings of pct0 for each cluster and category see below
}

#cutoff.method for categories enriched with input genes filtered by q < 0.01 and 0.833... > Fold change (FC) > 1.2
#NSC1a - Cell Process (748 genes):                  cutoff.method = "pct0" & pct0 = 0.3
#NSC1a - Signal Processing (748 genes):             cutoff.method = "fndr"
#NSC1b - Cell Process (990 genes):                  cutoff.method = "fndr"
#NSC1b - Signal Processing (990 genes):             cutoff.method = "fndr"
#NSC2a - Cell Process (1129 genes):                 cutoff.method = "fndr"
#NSC2a - Signal Processing (1129 genes):            cutoff.method = "fndr"
#NSC2b - Cell Process (628 genes):                  cutoff.method = "fndr"
#NSC2b - Signal Processing (628 genes):             cutoff.method = "fndr"
#Apoptotic NSC - Cell Process (347 genes):          cutoff.method = "pct0" & pct0 = 0.4
#Apoptotic NSC - Signal Processing (347 genes):     cutoff.method = "pct0" & pct0 = 0.2
#NCSC - Cell Process (2618 genes):                  cutoff.method = "fndr"
#NCSC - Signal Processing (2618 genes):             cutoff.method = "fndr"
#Apoptotic NCSC - Cell Process (517 genes):         cutoff.method = "pct0" & pct0 = 0.3
#Apoptotic NCSC - Cell Process (517 genes):         cutoff.method = "fndr"
#Glial precursor - Cell Process (3673 genes):       cutoff.method = "fndr"
#Glial precursor - Signal Processing (3673 genes):  cutoff.method = "fndr"
#Immature neurons - Cell Process (364 genes):       cutoff.method = "pct0" & pct0 = 0.7
#Immature neurons - Cell Process (364 genes):       cutoff.method = "fndr"

#cutoff.method for categories enriched with input genes filtered by q < 0.05
#NSC1a - Cell Process (3988 genes):                 cutoff.method = "fndr"
#NSC1a - Signal Processing (3988 genes):            cutoff.method = "fndr"
#NSC1b - Cell Process (4475 genes):                 cutoff.method = "fndr"
#NSC1b - Signal Processing (4475 genes):            cutoff.method = "fndr"
#NSC2a - Cell Process (3979 genes):                 cutoff.method = "fndr"
#NSC2a - Signal Processing (3979 genes):            cutoff.method = "fndr"
#NSC2b - Cell Process (2027 genes):                 cutoff.method = "fndr"
#NSC2b - Signal Processing (2027 genes):            cutoff.method = "fndr"
#Apoptotic NSC - Cell Process (551 genes):          cutoff.method = "fndr"
#Apoptotic NSC - Signal Processing (551 genes):     cutoff.method = "fndr"
#NCSC - Cell Process (3177 genes):                  cutoff.method = "fndr"
#NCSC - Signal Processing (3177 genes):             cutoff.method = "fndr"
#Apoptotic NCSC - Cell Process (586 genes):         cutoff.method = "fndr"
#Apoptotic NCSC - Cell Process (586 genes):         cutoff.method = "fndr"
#Glial precursor - Cell Process (4090 genes):       cutoff.method = "fndr"
#Glial precursor - Signal Processing (4090 genes):  cutoff.method = "fndr"
#Immature neurons - Cell Process (515 genes):       cutoff.method = "fndr"
#Immature neurons - Cell Process (515 genes):       cutoff.method = "fndr"

#fdr.out$qval #q-values to be stored for each category

corrected.pval <- cbind(fdr.out$pval,fdr.out$qval)
colnames(corrected.pval) <- c("p-value Fisher's exact test","q-value")
rownames(corrected.pval) <- signal_pathways.nsc1a
corrected.pval[corrected.pval[, "q-value"] < 0.05,]

